tests/integration/test_charm.py

```diff
@@ -0,0 +1,112 @@
+    """Returns a dict with GenericNamespacedResource as value for each CRD key."""
+    crds_kinds = [
+        crd["spec"]["names"]
+        for crd in yaml.safe_load_all(Path("./src/crds_manifests.yaml").read_text())
+    ]
+    jobs_classes = {}
+    for kind in crds_kinds:
+        job_class = lightkube.generic_resource.create_namespaced_resource(
+            group="kubeflow.org", version="v1", kind=kind["kind"], plural=kind["plural"]
+        )
+        jobs_classes[kind["kind"]] = job_class
+    return jobs_classes
+
+
+# TODO: Kubeflow upstream MXNet examples use GPU.
+# Not testing MXNetjobs until we have a CPU mxjob example.
+def test_create_training_jobs(ops_test: OpsTest):
+    """Validates that a training job can be created and is running.
+
+    Asserts on the *Job status.
+    """
+    namespace = ops_test.model_name
+    lightkube_client = lightkube.Client()
+    jobs_classes = lightkube_create_global_resources()
+
+    for p in glob.glob("examples/*.yaml"):
+        # Set up for creating an object of kind *Job
+        job_yaml = yaml.safe_load(Path(p).read_text())
+        job_object = lightkube.codecs.load_all_yaml(yaml.dump(job_yaml))[0]
+        job_class = jobs_classes[job_object.kind]
+
+        # Create *Job and check if it exists where expected
+        lightkube_client.create(job_object, namespace=namespace)
+        assert (
+            lightkube_client.get(
+                job_class, name=job_object.metadata.name, namespace=namespace
+            )
+            is not None
+        ), f"{job_object.metadata.name} does not exist"
+
+        # Wait for the *Job to have a status
+        # TODO: change this workaround after
+        # we have an implementation in lightkube
+        try:
+            lightkube_client.get(
+                job_class.Status, name=job_object.metadata.name, namespace=namespace
+            ).status["conditions"]
+        except TypeError:
+
+            @tenacity.retry(
+                wait=tenacity.wait_exponential(multiplier=20, min=1, max=60),
+                stop=tenacity.stop_after_delay(60),
+                reraise=True,
+            )
+            def get_job_status():
+                """Returns a lightkube object.
+
+                Retries multiple times using tenacity to allow time for the training job
+                to change its status from None -> Created -> Running/Succeeded.
+                """
+
+                return lightkube_client.get(
+                    job_class.Status, name=job_object.metadata.name, namespace=namespace
+                ).status["conditions"]
+
+            job_status = get_job_status()
+
+        # Check whether the *Job was created
+        assert (
+            job_status[0]["type"] == "Created"
+        ), f"Something went wrong creating {job_object.metadata.name}"
+
+        # Check wether the *Job is in a Running/Success state
+        assert (
+            job_status[-1]["type"] == "Running" or job_status[-1]["type"] == "Succeeded"
```

Comment:
If there was a short delay (a few seconds) between job creation and job running/succeeded, what would you like this test to do?  As is, it will fail, but I think you'd want it to succeed (a few second delay isn't a sign of something broken, could just be a random computer thing that resolves itself...).  Also, are you confident that any job entering the `Running` state confirms the operator is working correctly?  If yes then this works, but if no then I'd say just test for "Succeeded" since I think we're using tests cases that we expect to succeed.

I think this assertion would do better with the same `retry` pattern as above.  For some acceptable period (something long enough you know if it takes that long things have gone wrong) check for the acceptable statuses (`succeeded` and maybe `running`) and if you don't see them retry.  This will handle any race conditions or pauses, and if you're looking for `succeeded` rather than `running` it also waits until the job actually finishes.  

You can also add something that exits the checks if it hits a known failure status (eg: `if status == 'error', dont retry and instead fail`)

Reply:
Re status: I might have misunderstood the jobs.  I'd think they all have convergence or epoch limits that would end them but I might be wrong.  The pytorchjob at least looks like it will end.  I've never used xgboost though so not sure about that one.  Testing success sounds better than just running to me, but we can start with checking for either and see if it causes trouble.  

My hesitance comes from not knowing if the operator can still be broken after we see a `running` state (like if some data storage or I don't know what isn't wired properly), but let's just try it as is and see what happens :)

Reply:
Nvm @ca-scribner , you are right, eventually all jobs will have `Succeeded`. 
We may want to keep `Running` because:
1) `Succeeded` depends on so much things that an arbitrary wait may not allow to happen. Having different examples and setting a wait that covers all cases could be problematic. For reference, [Conditions].(https://www.kubeflow.org/docs/components/training/tftraining/#conditions)
2) "`TFJobRunning` means all sub-resources (e.g. services/pods) of this TFJob have been successfully scheduled and launched and the job is running." which sounds enough to me.

Reply:
sounds good to me too then, thanks!

Reply:
Resolving conversation.